contracts/transfer/src/state.rs

```diff
@@ -201,6 +201,60 @@ impl TransferState {
         self.mint_withdrawal("WITHDRAW", withdraw);
     }
 
+    /// Takes the deposit addressed to this contract, and immediately withdraws
+    /// it, effectively performing an atomic conversion between Phoenix notes
+    /// and Moonlight balance.
+    ///
+    /// This functions checks whether the deposit included with the transaction
+    /// is the exact value included in `convert`, and imposes that the
+    /// caller is indeed this contract.
+    ///
+    /// # Panics
+    /// This can only be called by this contract - the transfer contract - and
+    /// will panic if this is not the case.
+    pub fn convert(&mut self, convert: Withdraw) {
+        if rusk_abi::caller() != TRANSFER_CONTRACT {
+            panic!("Only a direct call can be a conversion");
+        }
+        if *convert.contract() != TRANSFER_CONTRACT.to_bytes() {
+            panic!("The conversion must target the transfer contract");
+        }
+
+        let deposit = transitory::deposit_info_mut();
+        match deposit {
+            Deposit::Available(_, deposit_value) => {
+                let deposit_value = *deposit_value;
+
+                if convert.value() != deposit_value {
+                    panic!("The value to convert doesn't match the value in the transaction");
+                }
+
+                // This is a direct contract call, as the first check in this
+                // function indicates, and the target of a deposit is always the
+                // direct contract called. Therefore, there is no need to check
+                // the available deposit targets this contract.
+                //
+                // if deposit_contract != TRANSFER_CONTRACT {
+                //     panic!();
+                // }
+
+                // Handle the withdrawal part of the conversion and set the
+                // deposit as being taken. Interesting to note is that we don't
+                // need to change the value held by the contract at all, since
+                // it never changes.
+                self.mint_withdrawal("CONVERT", convert);
+                *deposit = Deposit::Taken(TRANSFER_CONTRACT, deposit_value);
+            }
+            Deposit::None => panic!("There is no deposit in the transaction"),
+            // The transfer contract never calls this function directly, and the
+            // first check in this function is that the caller is the transfer
+            // contract. Therefore, the only way this code is reached is as a
+            // first contract call, and a deposit already being taken is
+            // impossible.
```

Comment:
I am confused - what does "first contract call" mean? Could you explain?

Reply:
I've yet to find the words to describe it properly. What is meant by "first contract call" is something like "the first execution of a contract besides `spend_and_execute`". To be more explicit, it's the call the maker of the transaction intends to make, and puts in the `exec` field of a transaction.

We've used "the direct contract" in the past as well. If you have suggestions for how to express it better I'm all ears(or eyes).

Reply:
So it is not a transfer contract self call and not an inter-contract call, but a `spend_and_execute` driven call. I understand, it is clear now, thanks. I don't know how to name it, but one more thing bothers me. Why is this used in reasoning about deposit? I can't see how the way it is called proves anything about deposit? Wouldn't it be more correct to say `deposit already being taken is not allowed`? I just don't think the reasoning and this comment should be written from the point of view of a person who knows what happens in the entire contract. It should be written from the position of knowledge of this method only IMO.

Reply:
To clarify even further, all transactions go through `spend_and_execute` regardless. This refers to the contract call in the transaction being to the transfer contract and that function explicitly, and not through some other contract.

That last point I disagree with. The contract should be, and can *only* be (unless there is no shared state), written from an imperative perspective. This is even enshrined, in a way, by the existence of `transitory`. If this wasn't the case transitory shouldn't exist.

The assumption made here is that there's only a single contract call in a transaction. Therefore, if we make the check in the code, we have a guarantee that it is the first contract call, so a deposit could never have been taken in the first place.

Reply:
> That last point I disagree with. The contract should be, and can _only_ be (unless there is no shared state), written from an imperative perspective. This is even enshrined, in a way, by the existence of `transitory`. If this wasn't the case transitory shouldn't/wouldn't exist.
> The assumption made here is that there's only a single contract call in a transaction. Therefore, if we make the check in the code, we have a guarantee that it is the first contract call, so a deposit could never have been taken in the first place.

I think you are commenting my comment which I deleted, because I got the exact point you are making by myself as well. So all good from my side.